utils/utils.go

```diff
@@ -438,39 +438,30 @@ func DefaultJsonPostProcess( apiResponseMap map[generic_structs.ComparableApiReq
     parsedErrorStructure := make(map[string]interface{})
 
     for request, response := range apiResponseMap {
-
-        // Catch JSON slices that don't have a map at the root
-        var jsonSlice []interface{}
-        err := json.Unmarshal(response, &jsonSlice)
-        if err == nil {
-            LogWarn("DefaultJsonPostProcess", "JSON is a slice - building map.",
-                err)
-            // Maybe more efficient, but less robust than build and marshal?
-            response = append( []byte("{\"items\":"),
-                append( response, []byte("}") ... ) ... )
-            // Add our new key we created to the base key expected.
-            for i, v := range jsonKeys {
-                if v["api_call_uuid"] == request.Uuid {
-                    length, err := strconv.Atoi(v["key_count"])
-                    if err != nil {
-                        LogFatal("DefaultJsonPostProcess",
-                            "Non-integer key_count is invalid", err)
-                    }
-                    for ci := 0; ci < length; ci++ {
-                        keyString := "current_base_key_" + strconv.Itoa(ci)
-                        if _, ok := jsonKeys[i][keyString]; ok ||
-                              jsonKeys[i][keyString] == "" {
-                            jsonKeys[i][keyString] = "items"
-                        } else {
-                            jsonKeys[i][keyString] = "items." +
-                                jsonKeys[i][keyString]
-                        }
+        // Maybe more efficient, but less robust than build and marshal?
+        response = append( []byte("{\"items\":"),
+            append( response, []byte("}") ... ) ... )
+        // Add our new key we created to the base key expected.
+        for i, v := range jsonKeys {
+            if v["api_call_uuid"] == request.Uuid {
+                length, err := strconv.Atoi(v["key_count"])
+                if err != nil {
+                    LogFatal("DefaultJsonPostProcess",
+                        "Non-integer key_count is invalid", err)
+                }
+                for ci := 0; ci < length; ci++ {
+                    keyString := "current_base_key_" + strconv.Itoa(ci)
+                    if _, ok := jsonKeys[i][keyString]; !ok || jsonKeys[i][keyString] == "" {
```

Comment:
You see how we're accounting for a blank string here when we're setting the slice of maps which include current_base_key's?  We just need to add this same check _outside_ of the if statement as well (so we handle the case of a blank `current_base_key` when the overall response _is_ a slice and when it _isn't_ a slice).

Reply:
@robwar Sharing the background of the issue.

This is the Expected Output with  GCP healthCheck get api  https://compute.googleapis.com/compute/v1/projects/{project}/global/healthChecks/{resourceId}
```
{
  "id": "2087338216676386660",
  "creationTimestamp": "2019-10-30T20:48:27.433-07:00",
  "name": "op-health-check",
  "description": "Checking the Health of operational resouces",
  "checkIntervalSec": 5,
  "timeoutSec": 5,
  "unhealthyThreshold": 2,
  "healthyThreshold": 2,
  "type": "TCP",
  "tcpHealthCheck": {
    "port": 3389,
    "request": "check",
    "response": "checkSuccessful",
    "proxyHeader": "NONE"
  },
  "selfLink": "https://www.googleapis.com/compute/v1/projects/saroj-dev/global/healthChecks/op-health-check",
  "kind": "compute#healthCheck"
}
```
 
 With the Resource available we used the epico library with the xml self-defined as follows 
 Method 1: 
```
  - name: "healthChecks1"
    vars:
      service: "health"
      project: "saroj-dev"
      resourceId: "op-health-check"
    endpoint: "https://compute.googleapis.com/compute/v1/projects/{{project}}/global/healthChecks/{{resourceId}}"
    current_base_key: ["id","creationTimestamp","name","description","tcpHealthCheck"]
    desired_base_key: ["healthCheck.id","healthCheck.creationTimestamp","healthCheck.name","healthCheck.description", "healthCheck.tcpHealthCheck"]
    current_error_key: ["error.errors","error.errors","error.errors","error.errors","error.errors"]
    desired_error_key: ["Errors.{{service}}","Errors.{{service}}","Errors.{{service}}","Errors.{{service}}","Errors.{{service}}"]
    documentation: "https://cloud.google.com/compute/docs/reference/rest/v1/healthChecks/list"
    params:
      querystring:
      header:
      body:
```
​
The Output that we get is 
```
"healthCheck": {
    "tcpHealthCheck": [{
      "request": "check",
      "proxyHeader": "NONE",
      "port": 3389,
      "response": "checkSuccessful"
    }],
    "creationTimestamp": ["2019-10-30T20:48:27.433-07:00"],
    "description": ["Checking the Health of operational resouces"],
    "name": ["op-health-check"],
    "id": ["2087338216676386660"]
  }
```
The problem with this response is we have to write down all the key value and in case of large size it will not be good enough. Also the output of each key is kept in JSONArray which is also not good as well.
​
Again we tried same response with another way possible
Method2:
```
  - name: "healthChecks2"
    vars:
      service: "health"
      project: "saroj-dev"
      resourceId: "op-health-check"
    endpoint: "https://compute.googleapis.com/compute/v1/projects/{{project}}/global/healthChecks/{{resourceId}}"
    current_base_key: ["id","creationTimestamp","name","description","tcpHealthCheck"]
    desired_base_key: ["healthCheck2","healthCheck2","healthCheck2","healthCheck2", "healthCheck2"]
    current_error_key: ["error.errors","error.errors","error.errors","error.errors","error.errors"]
    desired_error_key: ["Errors.{{service}}","Errors.{{service}}","Errors.{{service}}","Errors.{{service}}","Errors.{{service}}"]
    documentation: "https://cloud.google.com/compute/docs/reference/rest/v1/healthChecks/list"
    params:
      querystring:
      header:
      body:
```
The output we get here is
```
"healthCheck2": [
    "2087338216676386660",
    "2019-10-30T20:48:27.433-07:00",
    "op-health-check",
    "Checking the Health of operational resouces",
    {
      "request": "check",
      "proxyHeader": "NONE",
      "port": 3389,
      "response": "checkSuccessful"
    }
  ]
```
The problem here is also same as above we have to write down each and every key and will be problematic in case of lengthy JSON. Also, the main problem will be the key is unknown in this response so it is not useful and unrealistic.
​
Method3:
Here we also tried with the empty value as suggested.
```
  - name: "healthChecks3"
    vars:
      service: "health"
      project: "saroj-dev"
      resourceId: "op-health-check"
    endpoint: "https://compute.googleapis.com/compute/v1/projects/{{project}}/global/healthChecks/{{resourceId}}"
    current_base_key: [""]
    desired_base_key: ["healthCheck3"]
    current_error_key: ["error.errors"]
    desired_error_key: ["Errors.{{service}}"]
    documentation: "https://cloud.google.com/compute/docs/reference/rest/v1/healthChecks/list"
    params:
      querystring:
      header:
      body:
```
This produce empty output 
```
"healthCheck3": []
```
With any of the case above we couldn't get the expected JSON Object.
To Get at least of Complete JSON object we modified the utils method.
IF we could get the JSON object in any case we missed that would be the best option for us.

Reply:
@sarosejoshi - I understand the background here, and you are correct.  Methods 1 and 2 are definitely _not_ how it should work.

Method three _should_ work, but per my comment above, we are currently just missing a check for the blank base key outside of the if statement that checks for a slice.

In fact, looking closer - it seems like this should be refactored a bit.  We should end the `if err == nil {` statement after line 450.  This will move the `current_base_key` handling outside that conditional so it will _always_ be run - regardless of whether or not we had to convert the slice to a map.

I don't have my laptop setup, but try this and let me know if it gives the desired result where we can use method 3:

```suggestion lines 442 - 475
                            // Catch JSON slices that don't have a map at the root
        var jsonSlice []interface{}
        err := json.Unmarshal(response, &jsonSlice)
        if err == nil {
            LogWarn("DefaultJsonPostProcess", "JSON is a slice - building map.",
                err)
            // Maybe more efficient, but less robust than build and marshal?
            response = append( []byte("{\"items\":"),
                append( response, []byte("}") ... ) ... )
        }
        // Add our new key we created to the base key expected.
        for i, v := range jsonKeys {
            if v["api_call_uuid"] == request.Uuid {
                length, err := strconv.Atoi(v["key_count"])
                if err != nil {
                    LogFatal("DefaultJsonPostProcess",
                        "Non-integer key_count is invalid", err)
                }
                for ci := 0; ci < length; ci++ {
                    keyString := "current_base_key_" + strconv.Itoa(ci)
                    if _, ok := jsonKeys[i][keyString]; ok ||
                          jsonKeys[i][keyString] == "" {
                        jsonKeys[i][keyString] = "items"
                    } else {
                        jsonKeys[i][keyString] = "items." +
                            jsonKeys[i][keyString]
                    }
                }
                // Duplicated names aren't allowed, but do happen with sub-
                //    endpoints.  In which case, all other input fields
                //    like the current base key should be the same.
                break
            }
        }
```